### PR TITLE
Port #916 to v1alpha2

### PR DESCRIPTION
**Pull request by chuckha**
Signed-off-by: Chuck Ha <chuckh@vmware.com>

**What this PR does / why we need it**:
This PR forward ports #916 from v1a1 to v1a2

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Related to #913 

**Special notes for your reviewer**:

**Release note**:
```release-note
Adds retries to tag related calls to aws.
```

### CODE REVIEW

File: pkg/cloud/services/ec2/natgateways.go [hunk 1]
```diff
@@ -58,12 +59,15 @@ func (s *Service) reconcileNatGateways() error {
...
-			if err != nil {
+			if err := wait.WaitForWithRetryable(wait.NewBackoff(), func() (bool, error) {
+				if err := tags.Ensure(converters.TagsToMap(ngw.Tags), &tags.ApplyParams{
+					EC2Client:   s.scope.EC2,
+					BuildParams: s.getNatGatewayTagParams(*ngw.NatGatewayId),
+				}); err != nil {
+					return false, err
+				}
+				return true, nil
+			}, []string{awserrors.ResourceNotFound}); err != nil {
```
Comment (detiber) [inline pkg/cloud/services/ec2/natgateways.go]:
```suggestion
			}, awserrors.ResourceNotFound); err != nil {
```

File: pkg/cloud/services/ec2/natgateways.go [hunk 2]
```diff
@@ -160,23 +164,30 @@ func (s *Service) createNatGateway(subnetID string) (*ec2.NatGateway, error) {
...
+	var out *ec2.CreateNatGatewayOutput
+	if err := wait.WaitForWithRetryable(wait.NewBackoff(), func() (bool, error) {
+		if out, err = s.scope.EC2.CreateNatGateway(&ec2.CreateNatGatewayInput{
+			SubnetId:     aws.String(subnetID),
+			AllocationId: aws.String(ip),
+		}); err != nil {
+			return false, err
+		}
+		return true, nil
+	}, []string{awserrors.InvalidSubnet}); err != nil {
```
Comment (detiber) [inline pkg/cloud/services/ec2/natgateways.go]:
```suggestion
	}, awserrors.InvalidSubnet); err != nil {
```

File: pkg/cloud/services/ec2/natgateways.go [hunk 3]
```diff
@@ -160,23 +164,30 @@ func (s *Service) createNatGateway(subnetID string) (*ec2.NatGateway, error) {
...
-	if err := tags.Apply(applyTagsParams); err != nil {
+	if err := wait.WaitForWithRetryable(wait.NewBackoff(), func() (bool, error) {
+		if err := tags.Apply(&tags.ApplyParams{
+			EC2Client:   s.scope.EC2,
+			BuildParams: s.getNatGatewayTagParams(*out.NatGateway.NatGatewayId),
+		}); err != nil {
+			return false, err
+		}
+		return true, nil
+	}, []string{awserrors.ResourceNotFound}); err != nil {
```
Comment (detiber) [inline pkg/cloud/services/ec2/natgateways.go]:
```suggestion
	}, awserrors.ResourceNotFound); err != nil {
```

File: pkg/cloud/services/ec2/routetables.go [hunk 4]
```diff
@@ -92,7 +98,13 @@ func (s *Service) reconcileRouteTables() error {
...
 		}
 
-		if err := s.associateRouteTable(rt, sn.ID); err != nil {
+		if err := wait.WaitForWithRetryable(wait.NewBackoff(), func() (bool, error) {
+			if err := s.associateRouteTable(rt, sn.ID); err != nil {
+				return false, err
+			}
+			return true, nil
+		}, awserrors.RouteTableNotFound, awserrors.SubnetNotFound); err != nil {
+			record.Warnf(s.scope.Cluster, "FailedAssociateRouteTable", "Failed to associate managed RouteTable %q with Subnet %q: %v", rt.ID, sn.ID, err)
```
Comment (detiber) [inline pkg/cloud/services/ec2/routetables.go]: I think this failure might be reported in assocationRouteTable() now

File: pkg/cloud/services/ec2/vpc.go [hunk 5]
```diff
@@ -148,18 +158,27 @@ func (s *Service) createVPC() (*v1alpha2.VPCSpec, error) {
...
+	if err := wait.WaitForWithRetryable(wait.NewBackoff(), func() (bool, error) {
+		if err := tags.Apply(&tags.ApplyParams{
+			EC2Client:   s.scope.EC2,
+			BuildParams: tagParams,
+		}); err != nil {
+			return false, err
+		}
+		return true, nil
+	}, awserrors.VPCNotFound); err != nil {
+		record.Warnf(s.scope.Cluster, "FailedTagVPC", "Failed to tag managed VPC %q: %v", *out.Vpc.VpcId, err)
```
Comment (detiber) [inline pkg/cloud/services/ec2/vpc.go]: duplicate event here?

File: pkg/cloud/services/ec2/routetables.go (hunk 4 above)
Comment (ncdc) [reply pkg/cloud/services/ec2/routetables.go]: Correct - safe to remove from here

File: pkg/cloud/services/ec2/vpc.go [hunk 6]
```diff
@@ -59,21 +60,30 @@ func (s *Service) reconcileVPC() error {
...
 	}
 
+	// Make sure attributes are configured
+	if err := wait.WaitForWithRetryable(wait.NewBackoff(), func() (bool, error) {
+		if err := s.ensureManagedVPCAttributes(vpc); err != nil {
+			return false, err
+		}
+		return true, nil
+	}, awserrors.VPCNotFound); err != nil {
+		record.Warnf(s.scope.Cluster, "FailedSetVPCAttributes", "Failed to set managed VPC attributes for %q: %v", vpc.ID, err)
```
Comment (ncdc) [inline pkg/cloud/services/ec2/vpc.go]: Safe to remove - this is embedded in `s.ensureManagedVPCAttributes` now

File: pkg/cloud/services/ec2/vpc_test.go [hunk 7]
```diff
@@ -142,9 +142,10 @@ func TestReconcileVPC(t *testing.T) {
 					}, nil)
 
 				m.DescribeVpcAttribute(gomock.AssignableToTypeOf(&ec2.DescribeVpcAttributeInput{})).
-					DoAndReturn(describeVpcAttributeFalse).AnyTimes()
```
Comment (ncdc) [inline pkg/cloud/services/ec2/vpc_test.go]: I can't comment on line 117 above - do you need to remove `CidrBlock` to match https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/916/files#diff-0133c773261155a5c13f644de72eb616L117?

File: pkg/cloud/services/ec2/routetables.go (hunk 4 above)
Comment (chuckha) [reply pkg/cloud/services/ec2/routetables.go]: yep

**Comment by chuckha** [comment]
updated

**Comment by ncdc** [comment]
LGTM. @detiber want to give it another glance?

**Pull request merged**